src/imitation/data/types.py

```diff
@@ -109,15 +119,39 @@ def __post_init__(self):
                 "obs and acts must have same number of timesteps: "
                 f"{len(self.obs)} != {len(self.acts)}"
             )
-        if len(self.obs) == 0:
-            raise ValueError("Must have non-zero number of observations.")
 
         if self.infos is not None and len(self.infos) != len(self.obs):
             raise ValueError(
                 "obs and infos must have same number of timesteps: "
                 f"{len(self.obs)} != {len(self.infos)}"
             )
 
+    @overload
+    def __getitem__(self: T, key: slice) -> T:
+        pass
+
+    @overload
+    def __getitem__(self, key: int) -> Dict[str, np.ndarray]:
+        pass
+
+    def __getitem__(self, key):
+        """See TransitionsMinimal docstring for indexing and slicing semantics."""
+        # Extract items using `dataclasses.fields` + dict comprehension instead of using
+        # `dataclasses.asdict` because `asdict` expensively and undocumentedly
+        # deep-copies every numpy array value.
+        # See https://stackoverflow.com/a/52229565/1091722.
+        d = {f.name: getattr(self, f.name) for f in dataclasses.fields(self)}
+        d_item = {k: v[key] for k, v in d.items()}
+
+        if isinstance(key, slice):
+            # Return type is the same as this dataclass. Replace field value with
+            # slices.
+            return dataclasses.replace(self, **d_item)
+        else:
+            assert isinstance(key, int)
+            # Return type is a dictionary. Array values have no batch dimension.
+            return d_item
```

Comment:
Torch's default collate doesn't play well with our `infos` attribute, as it tries to join together all the infos dicts into a single infos dict, rather than giving a list of dicts as we want... struggling to figure out a good solution to this problem right now.

I thought about hardcoding a "infos" case in `__getitem__` so that Torch sees a length-1 list of info dict whenever it is currently seeing a info dict, because joining together the length-1 lists without recursive joining will result in what we want. But this doesn't actually work, because Torch in fact will recursive join, and has an internal sanity check that errors if the info dicts have different numbers of keys on different samples: 
https://github.com/pytorch/pytorch/blob/c38a5cba0d4b310b0e24fbf6fe0fa14fe2b23d3a/torch/utils/data/_utils/collate.py#L76-L83

To fix this, I think that we will need to provide a custom collate_fn that performs the `infos` case first, and then defers to `default_collate` for everything else.

Reply:
Hmm seems like custom collate is the best way forward. Do we actually need `infos` for training though? I could also imagine some wrapper that just drops infos (or I think it's already optional in some places?)

Reply:
The reason I added in `infos` in the first place was for a convoluted potential MineRL use where we might want to track which transitions belong to which trajectory name in our data stream.

If it turns out we don't need this in the future, I'd be fine with just removing `infos` from `TransitionsMInimal`.

> I think it's already otpional in some places?
It's a Optional field of `Trajectory`. However, it's required in `TransitionsMinimal` right now, even if it's just an empty dict for each timestep.